core/trino-main/src/main/java/io/trino/sql/dialect/trino/TypeConstraint.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.dialect.trino;
+
+import io.trino.spi.type.EmptyRowType;
+import io.trino.spi.type.MultisetType;
+import io.trino.spi.type.RowType;
+import io.trino.spi.type.Type;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import static io.trino.sql.newir.FormatOptions.isValidIdentifier;
+import static java.util.Objects.requireNonNull;
+
+public record TypeConstraint(Predicate<Type> constraint)
+{
+    // Intermediate result row type.
+    // Row without fields is supported and represented as EmptyRowType.
+    // If row fields are present, they must have valid unique names.
```

Comment:
Currently, the intermediate relation row type has field names, and the fields are referenced by name with the FieldSelection operation. It will be refactored so that the row type is anonymous, and the fields will be referenced by index with the FieldReference operation.

Explanation: 
The query program must work correctly with the Memo and Equivalence Classes. For that purpose, the intermediate relation type must be generic. Each operation in an Equivalence Class must derive exactly the same output type, because they all must be compatible with the downstream program.
Due to this limitation, field names in the intermediate row type aren't very useful. We must use generic sequential names, for example `f_1`, `f_2`, `f_3`... Using indexes would be more concise.